woocommerce/woocommerce-ios
# Product List: show product list with infinite scroll

Fixes #1262 
Fixes #1266 

Update release notes:


## Changes

Note for code reviewers: sorry about the size of this PR, ~200 diffs are on tests, and the major changes are listed below.

- Added `productsTabProductCellPlaceholderImage` to `UIImage` extension for the product icon shown on the products tab table view cell
- Moved `displayRatingPrompt` implementation in several tab view controllers to an extension so that the code can be reused
- Fixed typo from `hightlightTextColor` to `highlightTextColor`
- Created table view cell `ProductsTabProductTableViewCell` with pairing view model `ProductsTabProductViewModel`
  - `detailsAttributedString` has more complicated logic based on design as in #1266
  - Unit tests for variations and stock status
- Created `ProductsViewController` with a table view that renders a list of products with pull to refresh and infinite scroll
  - Followup spinner footer UI polish: #1281 
- Created `ProductsViewControllerStateCoordinator` to handle `ProductsViewController`'s state (`ProductsViewControllerState`) with two functions that could alter the view controller state:
  - `transitionToSyncingState`
  - `transitionToResultsUpdatedState`
  - Basic unit tests
- Showed the new `ProductsViewController` in `MainTabBarController`

## Testing

Prerequisite: the store has some products preferably in different statuses
- Log out from the store
- Log in to a store
- Go to `Products` tab --> if the store has products, they should be listed as design in #1266 
- If the site has many products (like Bananza), scroll to the bottom --> the next page should be loaded and the table view should be updated
- Tap on a product --> the product details should be shown and can go back to the products tab
- If you have other stores, switch store in `My store` tab > Settings
- Repeat the steps above for other stores with or without products

## Example screenshots

no products | results, portrait | results, landscape
-- | -- | --
![Simulator Screen Shot - iPhone Xs - 2019-09-09 at 13 53 08](https://user-images.githubusercontent.com/1945542/64506180-2aef4100-d309-11e9-8993-4746b9d38493.png) | ![Simulator Screen Shot - iPhone Xs - 2019-09-09 at 13 50 58](https://user-images.githubusercontent.com/1945542/64506186-317db880-d309-11e9-9f4d-1d9eeab457d5.png) | ![Simulator Screen Shot - iPhone Xs - 2019-09-09 at 13 51 05](https://user-images.githubusercontent.com/1945542/64506188-32164f00-d309-11e9-9dfd-848b7c378e06.png)
